docs/README-develop-best-practices.md

```diff
@@ -0,0 +1,385 @@
+# Develoment Best Practices for cisco_node_utils APIs.
+
+* [Overview](#overview)
+* [YAML Development Best Practices](#ydbp)
+* [Common Object Development Best Practices](#odbp)
+* [MiniTest Development Best Practices](#mdbp)
+
+## <a name="overview">Overview</a>
+
+This document is intended to assist in developing cisco_node_utils API's that are consistent with current best practices.
+
+
+## <a name="ydbp">YAML Development Best Practices</a>
+
+* [Y1](#yaml1): All yaml feature entries should be kept in alphabetical order.
+* [Y2](#yaml2): Use anchors where needed for `config_get` and `config_get_token` entries.
+* Y3: Avoid nested optional matches.
+* [Y4](#yaml4): Use the `_template` feature when getting/setting the same value at multiple levels.
+* [Y5](#yaml5): When possible include a `default_value` that represents the system default value and document properties that don't have a system default. **Use strings rather then symbols when applicable**.
+* [Y6](#yaml6): When possible, use the same `config_get` show command for all properties and document any anomalies.
+* [Y7](#yaml7): Use Key-value wildcards instead of Printf-style wildcards.
+* Y8: Use of the `show running all` command is encouraged over `show <feature>` commands since not all `show <feature>` commands behave in the same manner across different cisco platforms.
+
+
+
+## <a name="odbp">Common Object Development Best Practices</a>
+
+* [CO1](#co1): If your feature can be configured under the `global` and `non-global vrfs` make sure to account for this in your object design.
+* [CO2](#co2): Make use of the equality operator allowing proper `instance1 == instance2` checks in the minitests.
+* [CO3](#co3): Let `''` represent 'not configured at all' rather then `nil` unless the property has a default value.
+* [CO4](#co4): Make sure all new properites have a `getter`, `setter` and `default_getter` method.
+* [CO5](#co5): Use singleton-like design for resources that cannot have mulitple instances.
+
+## <a name="mdbp">MiniTest Development Best Practices</a>
+
+* MT1: Ensure that **all new API's** have minitest coverage.
+* [MT2](#mt2): Use appropriate `assert_foo` and `refute_foo` statements rather than `assert_equal`.
+* [MT3](#mt3): Do not hardcode interface names.
+* [MT4](#mt4): Make use of the `config` helper method for device configuration instead of `@device.cmd`.
+* [MT5](#mt5): Make use of the `assert_show_match` and `refute_show_match` helper methods to validate expected outcomes in the CLI instead of `@device.cmd("show...")`.
+
+
+
+## YAML Examples:
```

Comment:
General style suggestion - instead of having the details above and the examples below, the above bullet lists should serve just as a table of contents, and all of the actual content should be below. E.g.:

```markdown
## <a name="ydbp">YAML Development Best Practices</a>

* [Y1](#yaml1): All yaml feature entries should be kept in alphabetical order.
* [Y2](#yaml2): Use anchors where needed for `config_get` and `config_get_token` entries.
...
## YAML Best Practices

### <a name="yaml1">Y1: All yaml feature entries should be kept in alphabetical order.

Please keep all feature names in alphabetical order, and all options under a feature in alphabetical order as well. As YAML permits duplicate entries (in which case the last entry overrides any earlier entries), keeping a consistent order helps to prevent accidentally introducing such duplication.

...

### [Y2](#yaml2): Use anchors where needed for `config_get` and `config_get_token` entries.

<explanation of what this rule means>
<justification for this rule>
<example(s) of this rule>

```